File: C++/Sources/RSA.cpp
#include "RSA.h"

#include<iostream>

#include <cmath>
#include <chrono>
#include <ostream>

void RSA::setPrivateKeys() {
    setPrivateKey();
    setPrivateKey2();
}
void RSA::setPrivateKeys(long long privateKey, long long privateKey2) {
    this->privateKey = privateKey;
    setPrivateKey2(privateKey2);
}
void RSA::setPrivateKeys(int privateKey, int privateKey2) {
    this->privateKey = privateKey;
    setPrivateKey2(privateKey2);
}
void RSA::setPrivateKey() {
    privateKey = genRandPrime().getFirst();
}
void RSA::setPrivateKey2() {
    if (privateKey != 0) {
        Pair<int> privateKey2_h = genRandPrime();
        if (privateKey2_h.getFirst() == privateKey) {
            privateKey2_h = genRandPrimeNotEqual(privateKey2_h.getSecond());
        }
        privateKey2 = privateKey2_h.getFirst();
    }
}
void RSA::setPrivateKey2(long long privateKey2) {
    this->privateKey2 = privateKey2;
}
void RSA::setPrivateKey2(int privateKey2) {
    this->privateKey2 = privateKey2;
}
void RSA::setModulus() {
    if (privateKey != 0 && privateKey2 != 0) {
        modulus = privateKey * privateKey2;
    }
}

void RSA::setPublicKey() {
    if (privateKey != 0 && privateKey2 != 0) {
        long long privateKeyMin1 = privateKey - 1;
        long long privateKey2Min1 = privateKey2 - 1;
        long long privateKeysMultiplied = privateKeyMin1 * privateKey2Min1;

        Pair<int> primeLinePair = genRandPrimeLessThan(privateKeysMultiplied);

        while (gcd(primeLinePair.getFirst(), privateKeysMultiplied) != 1) {
            std::cout << primeLinePair.getFirst() << " "  << privateKeysMultiplied << " " << gcd(primeLinePair.getFirst(), privateKeysMultiplied) << std::endl;
            primeLinePair = genRandPrimeNotEqual(primeLinePair.getSecond());
        }

        this->publicKey = primeLinePair.getFirst();
    }
}
void RSA::setPublicKey(long long publicKey) {
    this->publicKey = publicKey;
}
void RSA::setPublicKey(int publicKey) {
    this->publicKey = publicKey;
}

RSA::RSA() : PublicKeyCryptosystem() {
    setPrivateKey();
    setPrivateKey2();
    setModulus();
    setPublicKey();
}
RSA::RSA(long long privateKey, long long privateKey2) : PublicKeyCryptosystem(privateKey) {
    setPrivateKey2(privateKey2);
    setModulus();
    setPublicKey();
}
RSA::RSA(int privateKey, int privateKey2) : PublicKeyCryptosystem(privateKey) {
    setPrivateKey2(privateKey2);
    setModulus();
    setPublicKey();
}
RSA::RSA(long long privateKey, long long privateKey2, long long publicKey) : PublicKeyCryptosystem(privateKey) {
    setPrivateKey2(privateKey2);
    setModulus();
    setPublicKey(publicKey);
}
RSA::RSA(int privateKey, int privateKey2, int publicKey) : PublicKeyCryptosystem(privateKey) {
    setPrivateKey2(privateKey2);
    setModulus();
    setPublicKey(publicKey);
}
RSA::~RSA() {}

long long RSA::getPrivateKey2() {
    return privateKey2;
}
long long RSA::encrypt(long long plainText) {
    return modPow(plainText, publicKey, modulus);
}
long long RSA::timedEncrypt(long long plainText, std::ostream& os, char delim) {
    int numBits1 = floor(log2(privateKey)) + 1;
    int numBits2 = floor(log2(privateKey2)) + 1;
    int avgNumBits = (numBits1 + numBits2) / 2;
    
    auto startTime = std::chrono::steady_clock::now();
    long long encryptedText = encrypt(plainText);
    auto endTime = std::chrono::steady_clock::now();

    os << avgNumBits << delim << std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime).count() << std::endl;

    return encryptedText;
}
long long RSA::decrypt(long long cipherText) {
    long long prime1 = privateKey - 1;
    long long prime2 = privateKey2 - 1;
    long long decryptionMod = prime1 * prime2;
    long long decryptionKey = modInverse(publicKey, decryptionMod);
    return modPow(cipherText, decryptionKey, modulus);
}
long long RSA::timedDecrypt(long long cipherText, std::ostream& os, char delim) {
    int numBits1 = floor(log2(privateKey)) + 1;
    int numBits2 = floor(log2(privateKey2)) + 1;
    int avgNumBits = (numBits1 + numBits2) / 2;
    
    auto startTime = std::chrono::steady_clock::now();
    long long decryptedText = decrypt(cipherText);
    auto endTime = std::chrono::steady_clock::now();

    os << avgNumBits << delim << std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime).count() << std::endl;

    return decryptedText;
}

File: Java/Classes/RSA.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.math.BigInteger;

public class RSA extends PublicKeyCryptosystem {
    protected BigInteger privateKey2;

    public RSA() {
        super();
        setPrivateKey();
        setPrivateKey2();
        setModulus();
        setPublicKey();
    }
    public RSA(BigInteger privateKey, BigInteger privateKey2) {
        super(privateKey);
        setPrivateKey2(privateKey2);
        setModulus();
        setPublicKey();
    }
    public RSA(int privateKey, int privateKey2) {
        super(privateKey);
        setPrivateKey2(privateKey2);
        setModulus();
        setPublicKey();
    }
    public RSA(BigInteger privateKey, BigInteger privateKey2, BigInteger publicKey) {
        super(privateKey);
        setPrivateKey2(privateKey2);
        setModulus();
        setPublicKey(publicKey);
    }
    public RSA(int privateKey, int privateKey2, int publicKey) {
        super(privateKey);
        setPrivateKey2(privateKey2);
        setModulus();
        setPublicKey(publicKey);
    }
    protected void setPrivateKeys() {
        setPrivateKey();
        setPrivateKey2();
    }
    protected void setPrivateKeys(BigInteger privateKey, BigInteger privateKey2) {
        setPrivateKey(privateKey);
        setPrivateKey2(privateKey2);
    }
    protected void setPrivateKeys(int privateKey, int privateKey2) {
        setPrivateKey(privateKey);
        setPrivateKey(privateKey2);
    }
    protected void setPrivateKey() {
        try {
            long privateKey_h = super.genRandPrime().getFirst().longValue();
            this.privateKey = new BigInteger(String.valueOf(privateKey_h));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    protected void setPrivateKey2() {
        if (privateKey != null) {
            try {
                Pair<Integer> privateKey2_h = super.genRandPrime();
                if (privateKey2_h.getFirst() == privateKey.longValue()) {
                    privateKey2_h = genRandPrimeNotEqual(privateKey2_h.getSecond());
                }
                this.privateKey2 = new BigInteger(String.valueOf(privateKey2_h.getFirst()));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
    protected void setPrivateKey2(BigInteger privateKey2) {
        this.privateKey2 = new BigInteger(privateKey2.toString());
    }
    protected void setPrivateKey2(int privateKey2) {
        this.privateKey2 = new BigInteger(String.valueOf(privateKey2));
    }
    protected void setModulus() {
        if(privateKey != null && privateKey2 != null)  {
            this.modulus = new BigInteger(privateKey.multiply(privateKey2).toString());
        }
    }

    protected void setPublicKey() {
        if (super.privateKey != null && this.privateKey2 != null) {
            // Set Public Key to a random prime less than (privateKey-1)(privateKey2-1)
            try {
                BigInteger privateKeyMin1 = BigInteger.valueOf(super.privateKey.longValue()).subtract(BigInteger.ONE);
                BigInteger privateKey2Min1 = BigInteger.valueOf(this.privateKey2.longValue()).subtract(BigInteger.ONE);
                BigInteger privateKeysMultiplied = privateKeyMin1.multiply(privateKey2Min1);

                long privateKeysMultipliedLong = privateKeysMultiplied.longValueExact();

                Pair<Integer> primeLinePair = genRandPrimeLessThan(privateKeysMultipliedLong);

                while (privateKeysMultiplied.gcd(BigInteger.valueOf(primeLinePair.getFirst())).compareTo(BigInteger.ONE) != 0) {
                    primeLinePair = genRandPrimeNotEqual(primeLinePair.getSecond());
                }

                this.publicKey = new BigInteger(String.valueOf(primeLinePair.getFirst()));

            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
    protected void setPublicKey(BigInteger publicKey) {
        this.publicKey = new BigInteger(publicKey.toString());
    }
    protected void setPublicKey(int publicKey) {
        this.publicKey = new BigInteger(String.valueOf(publicKey));
    }
    public BigInteger getPrivateKey2() {
        return privateKey2;
    }
    public BigInteger encrypt(BigInteger plainText) {
        BigInteger cipherText = plainText.modPow(publicKey, modulus);
        return cipherText;
    }
    public BigInteger timedEncrypt(BigInteger plainText, File output, Character delim) throws IOException {
        int bitLength1 = privateKey.bitLength();
        int bitLength2 = privateKey2.bitLength();
        double avgBitLength = (bitLength1 + bitLength2) / 2;

        long startTime = System.nanoTime();
        BigInteger encryptedText = encrypt(plainText);
        long endTime = System.nanoTime();

        output.createNewFile();
        FileWriter writer = new FileWriter(output.getName(), true);
        PrintWriter printer = new PrintWriter(writer);
        printer.println(avgBitLength + delim.toString() + (endTime - startTime));
        writer.close();
        printer.close();

        return encryptedText;
    }
    public BigInteger decrypt(BigInteger cipherText) {
        BigInteger prime1 = privateKey.subtract(BigInteger.ONE);
        BigInteger prime2 = privateKey2.subtract(BigInteger.ONE);
        BigInteger decryptionMod = prime1.multiply(prime2);
        try {
            BigInteger decryptionKey = publicKey.modInverse(decryptionMod);
            return cipherText.modPow(decryptionKey, modulus);
        } catch (ArithmeticException e) {
            System.out.println("Public key: " + publicKey);
            System.out.println("Decryption mod: " + decryptionMod);
            e.printStackTrace();
        }
        return null;
    }
    public BigInteger timedDecrypt(BigInteger cipherText, File output, Character delim) throws IOException {
        int bitLength1 = privateKey.bitLength();
        int bitLength2 = privateKey2.bitLength();
        double avgBitLength = (bitLength1 + bitLength2) / 2;

        long startTime = System.nanoTime();
        BigInteger decryptedText = decrypt(cipherText);
        long endTime = System.nanoTime();

        output.createNewFile();
        FileWriter writer = new FileWriter(output.getName(), true);
        PrintWriter printer = new PrintWriter(writer);
        printer.println(avgBitLength + delim.toString() + (endTime - startTime));
        writer.close();
        printer.close();

        return decryptedText;
    }
}
